incubator/hnc/pkg/validators/object.go

```diff
@@ -0,0 +1,117 @@
+package validators
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	api "github.com/kubernetes-sigs/multi-tenancy/incubator/hnc/api/v1alpha1"
+	"github.com/kubernetes-sigs/multi-tenancy/incubator/hnc/pkg/forest"
+)
+
+// ObjectsServingPath is where the validator will run. Must be kept in sync with the
+// kubebuilder markers below.
+const (
+	ObjectsServingPath = "/validate-objects"
+)
+
+// Note: the validating webhook FAILS OPEN. This means that if the webhook goes down, all further
```

Comment:
Do we want this? I'd think failing closed would be more ideal here (for better or worse)?

Reply:
@adrianludwin  thoughts?

Reply:
For now, yes, I think so - if the HNC (which is pre-alpha!) crashes, and the webhook fails closed, _all_ modifications to any of these objects will be rejected. Maybe one day we can change this. When I spoke to @liggitt he felt quite strongly about this.

By contrast, changes to the HierarchyConfiguration _do_ fail closed because the worst that can happen is that you can no longer modify your (existing, presumably valid) configuration.

Reply:
> if the HNC (which is pre-alpha!) crashes, and the webhook fails closed, all modifications to any of these objects will be rejected.

I'd argue if you expect a pre-alpha thing not to affect your cluster it's on you. My main concern is the controller going down for a short time and then uses being allowed to circumvent the HNC. This is less a concern for HNC but more when we wrap a Tenant controller around it.

If you're deploying HNC to  your cluster to have this hierarchy, you're likely relying on it to provide some form access controls for your users. If HNC goes down, then maybe it's ok if all objects are rejected? Or at least it's configurable. It really depends on the usecase. If HNC is just to help sync across namespaces it won't matter that much, but if you are trying to achieve some form of isolation from this could have negative consequences.

Regardless, we don't have to figure this out now. I would like to revisit this at some point though.

Reply:
I mean, it is configurable by editing the manifest 😁  This is just the default.

Reply:
With that said, I filed #188 to revisit this default in v0.2. Thanks!

Reply:
Good point. I will leave it for a future PR. This PR will focus on implementation of the validator.